iroh-net/src/magicsock.rs

```diff
@@ -2518,56 +2420,97 @@ fn bind(
 #[derive(derive_more::Debug, Default, Clone)]
 struct DiscoveredDirectAddrs {
     /// The last set of discovered direct addresses.
-    addrs: Vec<DirectAddr>,
+    addrs: Watchable<BTreeSet<DirectAddr>>,
 
     /// The last time the direct addresses were updated, even if there was no change.
-    updated_at: Option<Instant>,
-}
-
-impl PartialEq for DiscoveredDirectAddrs {
-    fn eq(&self, other: &Self) -> bool {
-        endpoint_sets_equal(&self.addrs, &other.addrs)
-    }
+    ///
+    /// This is only ever None at startup.
+    updated_at: Arc<RwLock<Option<Instant>>>,
 }
 
 impl DiscoveredDirectAddrs {
-    fn new(endpoints: Vec<DirectAddr>) -> Self {
-        Self {
-            addrs: endpoints,
-            updated_at: Some(Instant::now()),
+    /// Updates the direct addresses, returns `true` if they changed, `false` if not.
+    fn update(&self, addrs: BTreeSet<DirectAddr>) -> bool {
+        *self.updated_at.write().unwrap() = Some(Instant::now());
+        let updated = self.addrs.update(addrs).is_ok();
+        if updated {
+            event!(
+                target: "events.net.direct_addrs",
+                Level::DEBUG,
+                addrs = ?self.addrs.get(),
+            );
         }
+        updated
     }
 
-    fn into_iter(self) -> impl Iterator<Item = DirectAddr> {
-        self.addrs.into_iter()
+    fn last_refresh_ago(&self) -> Option<Duration> {
+        self.updated_at
+            .read()
+            .expect("poisoned")
+            .map(|at| at.elapsed())
     }
 
-    fn iter(&self) -> impl Iterator<Item = &DirectAddr> + '_ {
-        self.addrs.iter()
-    }
-
-    fn is_empty(&self) -> bool {
-        self.addrs.is_empty()
+    fn sockaddrs(&self) -> BTreeSet<SocketAddr> {
+        self.addrs.read().iter().map(|da| da.addr).collect()
     }
 
     fn fresh_enough(&self) -> bool {
-        match self.updated_at.as_ref() {
+        match *self.updated_at.read().expect("poisoned") {
             None => false,
             Some(time) => time.elapsed() <= ENDPOINTS_FRESH_ENOUGH_DURATION,
         }
```

Comment:
```suggestion
        self.last_refresh_ago().map_or(false, |dur| dur <= ENDPOINTS_FRESH_ENOUGH_DURATION)
```

Reply:
TIL about `Option::map_or`.  I guess it's shorter than `.map().unwrap_or()`, not yet convinced I like it.

Anyway, this is subjective.  I don't have anything against this suggestion but also don't particularly like it.  Probably because I'm not really happy with the `last_refresh_ago` name...

Reply:
it's barely english, so it's understandable that you aren't happy with it 😂